apps/linear_algebra/src/blas_l3_generators.cpp

```diff
@@ -42,64 +42,152 @@ class GEMMGenerator :
         }
     }
 
+    Func transpose(ImageParam im) {
+        Func transpose_tmp("transpose_tmp"), im_t("im_t");
+        Var i("i"), j("j"), ii("ii"), ji("ji"),
+            ti("ti"), tj("tj"), t("t");
+
+        transpose_tmp(i, j) = im(j, i);
+        im_t(i, j) = transpose_tmp(i, j);
+
+        Expr rows = im.width(), cols = im.height();
+
+        im_t.compute_root()
+            .specialize(rows >= 4 && cols >= 4)
+            .tile(i, j, ii, ji, 4, 4).vectorize(ii).unroll(ji)
+            .specialize(rows >= 128 && cols >= 128)
+            .tile(i, j, ti, tj, i, j, 16, 16)
+            .fuse(ti, tj, t).parallel(t);
+
+        transpose_tmp.compute_at(im_t, i)
+            .specialize(rows >= 4 && cols >= 4).vectorize(j).unroll(i);
```

Comment:
It's the vector size for a in-register transpose, so the complexity of the code is more related to the number of things to transpose than their bit width. Four seemed good for floats and doubles.